File: StructAndAlgLab3.cpp
#include <iostream>
#include <queue>
#include <string>

using namespace std;






struct Task
{
    int id;
    string description;
    int priority;
};









int inputNumber(string message)
{
    int number;

    cout << message;

    while (!(cin >> number))
    {
        cout << "Invalid input. Enter a number: ";

        cin.clear();
        cin.ignore(1000, '\n');
    }

    return number;
}









class TaskQueue
{
private:
    queue<Task> tasks;

public:
    void enqueue(Task task)
    {
        tasks.push(task);
    }

    void dequeue()
    {
        if (isEmpty())
        {
            cout << "Queue is empty\n";
            return;
        }

        tasks.pop();
        cout << "Task removed\n";
    }

    Task front()
    {
        return tasks.front();
    }

    bool isEmpty()
    {
        return tasks.empty();
    }

    string toString()
    {
        if (isEmpty())
        {
            return "Queue is empty\n";
        }

        queue<Task> temp = tasks;
        string result;

        while (!temp.empty())
        {
            Task current = temp.front();

            result += "ID: " + to_string(current.id);
            result += ", Description: " + current.description;
            result += ", Priority: " + to_string(current.priority);
            result += "\n";

            temp.pop();
        }

        return result;
    }
};










int main()
{
    TaskQueue q;
    int choice;

    do
    {
        cout << "\n===== MENU =====\n";
        cout << "1. Add task\n";
        cout << "2. Remove first task\n";
        cout << "3. Show first task\n";
        cout << "4. Show queue\n";
        cout << "0. Exit\n";

        choice = inputNumber("Your choice: ");

        switch (choice)
        {
        case 1:
        {
            Task task;

            task.id = inputNumber("Enter ID: ");

            cin.ignore(1000, '\n');

            cout << "Enter description: ";
            getline(cin, task.description);

            task.priority = inputNumber("Enter priority: ");

            q.enqueue(task);

            cout << "Task added\n";
            break;
        }

        case 2:
            q.dequeue();
            break;

        case 3:
            if (q.isEmpty())
            {
                cout << "Queue is empty\n";
            }
            else
            {
                Task first = q.front();

                cout << "ID: " << first.id << endl;
                cout << "Description: " << first.description << endl;
                cout << "Priority: " << first.priority << endl;
            }
            break;

        case 4:
            cout << "\nQueue contents:\n";
            cout << q.toString();
            break;

        case 0:
            cout << "Program terminated\n";
            break;

        default:
            cout << "Invalid menu item\n";
        }

    } while (choice != 0);

    return 0;
}

File: StructAndAlgLab3.py
class Task:
    def __init__(self, task_id, description, priority):
        self.id = task_id
        self.description = description
        self.priority = priority

    def __str__(self):
        return f"id={self.id}, описание='{self.description}', приоритет={self.priority}"


class TaskQueue:
    def __init__(self):
        self.items = []

    def enqueue(self, task):
        self.items.append(task)

    def dequeue(self):
        if self.is_empty():
            print("Очередь пуста")
            return None
        return self.items.pop(0)

    def front(self):
        if self.is_empty():
            print("Очередь пуста")
            return None
        return self.items[0]

    def is_empty(self):
        return len(self.items) == 0

    def __str__(self):
        if self.is_empty():
            return "Очередь пуста"
        parts = " -> ".join(f"[{task}]" for task in self.items)
        return f"Очередь: {parts}"


queue = TaskQueue()

while True:
    print("\n1 — Добавить задачу")
    print("2 — Извлечь задачу (dequeue)")
    print("3 — Посмотреть первую (front)")
    print("4 — Проверить isEmpty")
    print("5 — Вывести очередь")
    print("0 — Выход")

    choice = input("Выберите действие: ").strip()

    if choice == "1":
        task_id_raw = input("ID задачи: ").strip()
        if not task_id_raw.isdigit():
            print("ID должен быть числом.")
        else:
            description = input("Описание: ").strip()
            priority_raw = input("Приоритет (число): ").strip()
            if not priority_raw.isdigit():
                print("Приоритет должен быть числом.")
            else:
                queue.enqueue(Task(int(task_id_raw), description, int(priority_raw)))
                print("Задача добавлена.")

    elif choice == "2":
        task = queue.dequeue()
        if task:
            print("Извлечена:", task)

    elif choice == "3":
        task = queue.front()
        if task:
            print("Первая задача:", task)

    elif choice == "4":
        print("Очередь пуста?", queue.is_empty())

    elif choice == "5":
        print(queue)

    elif choice == "0":
        print("До свидания!")
        break

    else:
        print("Неизвестная команда.")
